contracts/contracts/coordination/SigningCoordinator.sol

```diff
@@ -0,0 +1,377 @@
+// SPDX-License-Identifier: AGPL-3.0-or-later
+
+pragma solidity ^0.8.0;
+
+import "@openzeppelin/contracts/utils/cryptography/ECDSA.sol";
+import "@openzeppelin/contracts/utils/cryptography/MessageHashUtils.sol";
+import "@openzeppelin-upgradeable/contracts/access/extensions/AccessControlDefaultAdminRulesUpgradeable.sol";
+import "@openzeppelin-upgradeable/contracts/proxy/utils/Initializable.sol";
+import "./ISigningCoordinatorChild.sol";
+import "./ThresholdSigningMultisigCloneFactory.sol";
+import "./SigningCoordinatorDispatcher.sol";
+import "../TACoApplication.sol";
+
+// SigningCoordinator ----> Dispatcher ----> (Relevant) L1Sender ---------[BRIDGE]---------- L2Receiver ----> SigningCoordinatorChild (1. deploys multisig OR 2. updates multisig)
+
+contract SigningCoordinator is Initializable, AccessControlDefaultAdminRulesUpgradeable {
+    using MessageHashUtils for bytes32;
+    using ECDSA for bytes32;
+
+    struct SigningCohortParticipant {
+        address provider;
+        address operator;
+        bytes signature;
+    }
+
+    struct SigningCohort {
+        address initiator;
+        uint32 initTimestamp;
+        uint32 endTimestamp;
+        // TODO: what's the point of authority if we have a single global cohort?
+        address authority;
+        uint16 totalSignatures;
+        uint16 numSigners;
+        uint16 threshold;
+        SigningCohortParticipant[] signers;
+        uint256[] chains;
+        mapping(uint256 => bytes) conditions; // TODO: chainId -> condition itself or hash(condition)
+    }
+
+    bytes32 public constant INITIATOR_ROLE = keccak256("INITIATOR_ROLE");
+
+    mapping(uint32 => SigningCohort) public signingCohorts;
+    uint256 public numberOfSigningCohorts;
+
+    event InitiateSigningCohort(
+        uint32 indexed cohortId,
+        uint256 chainId,
+        address indexed authority,
+        address[] participants
+    );
+    event SigningCohortSignaturePosted(
+        uint32 indexed cohortId,
+        address indexed provider,
+        bytes signature
+    );
+    event SigningCohortDeployed(uint32 indexed cohortId, uint256 chainId);
+    event SigningCohortConditionsSet(uint32 indexed cohortId, uint256 chainId, bytes conditions);
+
+    event TimeoutChanged(uint32 oldTimeout, uint32 newTimeout);
+    event MaxCohortSizeChanged(uint16 oldSize, uint16 newSize);
+    event DispatcherChanged(address oldDispatcher, address newDispatcher);
+
+    enum SigningCohortState {
+        NON_INITIATED,
+        AWAITING_SIGNATURES,
+        TIMEOUT,
+        ACTIVE,
+        EXPIRED
+    }
+
+    TACoApplication public immutable application;
+    SigningCoordinatorDispatcher public signingCoordinatorDispatcher;
+
+    uint96 private immutable minAuthorization;
+
+    uint32 public timeout;
+    uint16 public maxCohortSize;
+
+    SigningCohortParticipant internal __sentinelSigner;
```

Comment:
just for unification - adding gap (similar as in Coordinator) will help in case any contract will inherit this one

Reply:
there is no strict rules, it could be any number, we usually start from 20

Reply:
looks good